man/fieldify.1
.TH FIELDIFY "1" "October 2018" "fieldify 0.1.0" "User Commands"

.SH NAME
fieldify - normalize field separators in streaming data

.SH SYNOPSIS
fieldify [\-h] [\-\-version] [\-\-ofs OFS] [\-\-irs IRS] [\-\-ors ORS]
[\-\-regex REGEX] [\-\-bufsize BUFSIZE]

.SH DESCRIPTION

.PP
This script consumes streaming data from standard in and splits it along a
regular expression, then replaces the splits with a set field separator. By
default, it collapses whitespace\-separated columns into tab\-separated columns.
.SS "optional arguments:"
.TP
\fB\-h\fR, \fB\-\-help\fR
show this help message and exit
.TP
\fB\-\-version\fR
show program's version number and exit
.TP
\fB\-\-ofs\fR OFS, \fB\-F\fR OFS
Specify output field separator. (default: \et)
.TP
\fB\-\-irs\fR IRS, \fB\-f\fR IRS
Override default input record separator. (default: \en)
.TP
\fB\-\-ors\fR ORS, \fB\-R\fR ORS
Override default output record separator. (default:
\en)
.TP
\fB\-\-regex\fR REGEX, \fB\-r\fR REGEX
Specify regex to split with, in Python 3.X re
format.This is used in lieu of an IFS. (default: \es
.TP
\fB\-\-bufsize\fR BUFSIZE, \fB\-b\fR BUFSIZE
Override read buffer size. (default: 4096)
.TP
\fB\-\-nomatchdrop\fR, \fB\-m\fR
By default, fields which match \fB\-\-regex\fR are dropped.
This prevents instances of the field separator from
appearing in the output. Asserting this option will
disable dropping such fields. This flag and the field
dropping behavior were added in 0.1.0.

.SH EXAMPLES

.PP
$ echo 'axxxbxczdxexxf' | fieldify --regex '[x]' --irs z

.IP
a	b	c
.br
d	e	f

.PP
$ echo 'a  b c\\nd    e       f' | fieldify

.IP
a	b	c
.br
d	e	f

.PP
$ echo 'a  b c\\nd    e       f' | fieldify --ofs ','

.IP
a,b,c
.br
d,e,f

.PP
$ echo 'a  b c\\nd    e       f' | fieldify --ofs ',' --ors '|'

.IP
a,b,c|d,e,f

.SH NOTES

.PP
The \fB--bufsize\fR flag should never affect the output, but it may be
beneficial to tweak it for performance reasons in different use cases. If you
observe different output while changing the buffer size, you should report this
as a bug.

.PP
The behavior of the --nomatchdrop flag is demonstrated below:

.PP
$ echo 'list, of, several, things, and a final thing' | fieldify -r '(, )|(, and )' --ofs '|'

.IP
list|of|several|things|and a final thing

.PP
$ echo 'list, of, several, things, and a final thing' | fieldify -r '(, )|(, and )' --ofs '|' --nomatchdrop

.IP
list|, |of|, |several|, |things|, |and a final thing

.SH CHANGELOG

.IP 0.0.1
Initial version.

.IP 0.1.0

.IP \bu 2
Fixed a bug when a regex was passed via --regex that caused re.split() to
return instances of None.

.IP \bu 2
Fields yielded by splitting on --regex which match the regex supplied via
--regex are now dropped. This avoids edge cases where field delimiters can
occasionally appear in the output.

.IP \bu 2
Implemented the --nomatchdrop flag.


.SH HISTORY
.PP
Fieldify began as a simple awk script, which behaved identically to fieldify
with the default flags. It was rewritten in September 2018 in Python to support
arbitrary field and record separators, and to allow the input record separator
to be specified as a regular expression.

.SH COPYRIGHT

(c) 2018 Charles A. Daniels, All Rights Reserved.


src/fieldify/fieldify.py
#!/usr/bin/env python3

# Copyright 2018 Charles Daniels

#  Redistribution and use in source and binary forms, with or without
#  modification, are permitted provided that the following conditions are met:

#  1. Redistributions of source code must retain the above copyright notice,
#  this list of conditions and the following disclaimer.

#  2. Redistributions in binary form must reproduce the above copyright notice,
#  this list of conditions and the following disclaimer in the documentation
#  and/or other materials provided with the distribution.

#  3. Neither the name of the copyright holder nor the names of its
#  contributors may be used to endorse or promote products derived from this
#  software without specific prior written permission.

#  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
#  AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
#  IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
#  ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
#  LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
#  CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
#  SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
#  INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
#  CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
#  ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
#  POSSIBILITY OF SUCH DAMAGE.


import argparse
import re
import sys

# when this flag is asserted, format_record() does not drop fields that match
# the record separator regex.
nomatchdrop = False

descr = """
This script consumes streaming data from standard in and splits it along a
regular expression, then replaces the splits with a set field separator.  By
default, it collapses whitespace-separated columns into tab-separated columns.
"""

version = "0.1.0"

def format_record(rec, r, ofs):
    """format_record

    Format the output of an entire record. The caller is expected to ensure
    that the partial records are not passed in.

    :param rec: record to format
    :param r: compiled regex to use for splitting in lieu of IFS
    :param ofs: output field separator
    """

    # drop fields that are empty or invalid
    fields = []
    for field in r.split(rec):
        if field is None:
            continue

        if field is "":
            continue

        if (r.match(field) and (not nomatchdrop)):
            # avoid edge cases where the field separator winds up in the split
            # output
            continue

        fields.append(field)

    return ofs.join(fields)

def extract_records(s, irs):
    """extract_records

    Extract records from the string s along the irs delimiter, returning a
    tuple of the record list and any remainder. If irs is not present in s,
    then this function returns ([], s). The caller is expected to grantee that
    s is non-empty.

    :param s:
    :param irs:
    """

    if irs not in s:
        return ([], s)
    else:
        records = s.split(irs)
        return (records[:-1], records[-1])

def main():

    parser = argparse.ArgumentParser(description=descr)

    parser.add_argument('--version', action='version', version=version)

    parser.add_argument("--ofs", "-F", default='\t',
            help="Specify output field separator. (default: \\t)")

    parser.add_argument("--irs", "-f", default='\n',
            help="Override default input record separator. (default: \\n)")

    parser.add_argument("--ors", "-R", default='\n',
            help="Override default output record separator. (default: \\n)")

    parser.add_argument("--regex", "-r", default='\s',
            help="Specify regex to split with, in Python 3.X re format." +
            "This is used in lieu of an IFS. (default: \\s")

    parser.add_argument("--bufsize", "-b", default=4096, type=int,
            help="Override read buffer size. (default: 4096)")

    parser.add_argument("--nomatchdrop", "-m", default=False,
            action="store_true", help="By default, fields which match " +
            "--regex are dropped. This prevents instances of the field " +
            "separator from appearing in the output. Asserting this option " +
            "will disable dropping such fields. This flag and the field " +
            "dropping behavior were added in 0.1.0.")

    args = parser.parse_args()

    bufsize = args.bufsize
    buf = ""
    ofs = args.ofs
    regex = args.regex
    irs = args.irs
    ors = args.ors
    r = None

    global nomatchdrop
    nomatchdrop=args.nomatchdrop

    try:
        r = re.compile(regex)
    except Exception as e:
        sys.stderr.write("ERROR: Invalid regex: {}\n".format(e))
        sys.exit(1)

    if bufsize < 16:
        sys.stderr.write("bufsize too small (< 16).\n")
        sys.exit(1)

    while True:
        buf_step = sys.stdin.read(bufsize)
        buf += buf_step

        if buf_step == '':
            # we are done reading
            if buf != '':
                # there is still text in the buffer
                sys.stdout.write(ors)
                sys.stdout.write(format_record(buf, r, ofs))
            break

        (records, buf) = extract_records(buf, irs)

        if len(records) == 0:
            continue

        for i in range(len(records)):
            sys.stdout.write(format_record(records[i], r, ofs))

            if i + 1 < len(records):
                # avoid emitting a spurious trailing ors
                sys.stdout.write(ors)


if __name__ == "__main__":
    main()
